pkg/statistics/handle/ddl/ddl.go

```diff
@@ -195,6 +200,64 @@ func (h *ddlHandlerImpl) HandleDDLEvent(t *util.DDLEvent) error {
 	return nil
 }
 
+// updateStatsWithCountDeltaAndModifyCountDelta updates
+// the global stats with the given count delta and modify count delta.
+// Only used by some special DDLs, such as exchange partition.
+func updateStatsWithCountDeltaAndModifyCountDelta(
+	sctx sessionctx.Context,
+	tableID int64,
+	countDelta, modifyCountDelta int64,
+) error {
+	lockedTables, err := lockstats.QueryLockedTables(sctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	isLocked := false
+	if _, ok := lockedTables[tableID]; ok {
+		isLocked = true
+	}
+	startTS, err := util.GetStartTS(sctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if isLocked {
+		// For locked tables, it is possible that the record gets deleted. So it can be negative.
+		_, err = util.Exec(
+			sctx,
+			"INSERT INTO mysql.stats_table_locked "+
+				"(version, count, modify_count, table_id) "+
+				"VALUES (%?, %?, %?, %?) "+
+				"ON DUPLICATE KEY UPDATE "+
+				"version = VALUES(version), "+
+				"count = count + VALUES(count), "+
+				"modify_count = modify_count + VALUES(modify_count)",
+			startTS,
+			countDelta,
+			modifyCountDelta,
+			tableID,
+		)
+		return err
+	}
+	// Because count can not be negative, so we need to get the current and calculate the delta.
+	_, err = util.Exec(
+		sctx,
+		"INSERT INTO mysql.stats_meta "+
+			"(version, count, modify_count, table_id) "+
+			"SELECT %?, GREATEST(0, count + %?), GREATEST(0, modify_count + %?), %? "+
+			"FROM mysql.stats_meta WHERE table_id = %? "+
+			"ON DUPLICATE KEY UPDATE "+
+			"version = VALUES(version), "+
+			"count = VALUES(count), "+
+			"modify_count = VALUES(modify_count)",
+		startTS,
+		countDelta,
+		modifyCountDelta,
+		tableID,
+		tableID,
```

Comment:
This SQL seems questionable.
The "insert" is actually missing, and the "update" is wrong.

Reply:
This works for updates, but not for inserts.